### PR TITLE
Add a second CUDA transform reduction strategy for small reduction sizes

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This revision adds a new strategy for the small reduction domain.
For small reductions, we rely on L1 cache locality and do not try to parallelize individual reductions: every thread performs the full reduction locally.

This is empirically determined profitable up until a reduction size of 64.
This can be fine-tuned later when we have better support for filtering and profitability considerations.

Running the iree-samples/transform_dialect script:
```
(  benchmark-transform-create -r  ../iree-samples/transform_dialect/benchmark_linalg_reductions.stub.mlir   reduction_2d_static 1234567 15 )
```

before:
```
With transform dialect: reduction_2d_static --function_input="1234567x15xf32" P50: 1278300.0000 ns 14.48682234217319877962 GElements/s
Without transform dialect: reduction_2d_static --function_input="1234567x15xf32" P50: 2778900.0000 ns 6.66396955629925510093 GElements/s
```

with this PR:
```
With transform dialect: reduction_2d_static --function_input="1234567x15xf32" P50: 150060.00 ns 123.40733706517393042782 GElements/s
Without transform dialect: reduction_2d_static --function_input="1234567x15xf32" P50: 2808400.0000 ns 6.59396987608602763139 GElements/s
```

### CODE REVIEW

**Review by matthias-springer — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:9]: spurious

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:24]: spurious

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:321]: Add a note: the ability of this strategy to run at peak is highly dependent on properly using vector<4/2>, this currently happens only on size multiples of 4/2. This has been corroborated by looking at perf dips on uneven sizes and inspecting the ptx.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:382]: ... because we never know we are in a small-reduction regime mode.
Since this mode does not coalesce reads, perf will often suffer badly.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:330]: without large efficiency

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:302]: This comment looks dangling. Maybe just say there's nothing to compute for stage 3.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:392]: Nit: prefer `isDynamicDim`.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:415]: It feels that these checks should become part of the matcher. A dimension being static and being less than X were the primary use cases for them.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:645]: I'm not a fan of this. This makes the strategy too dynamic and difficult to express in the transform IR without adding control flow there, which we want to avoid. I'd rather play with empty handles.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:415]: Happy to evolve towards that but there are a few caveats right now:
  1. the first check is due to a compiler bug, that should be fixed and should not pollute the matchers.
  2. the other 2 are performance-related constraints for which we lack infra.

I'd prefer to avoid mixing constraints related to structural checks and capture of the IR (what we have today)
with additional constraints related to selecting the strategy and its parameters, which are also dependent on characteristics of the HW.

It will be great to use a similar mechanism for everything but I don't think we have that spelled out properly yet.

My intuition is your comment is more about setting up the stage towards that evolution rather than requesting a change right now?

Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:415]: will add TODOs

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:645]: ok, fair enough, I'll refactor calls to "TakeFirstOp" which for some reason I always have troubles with in my mental model ..

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (ftynse) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp]: Yes, I saw the compiler bug, that is not a problem. I find it really, really unfortunate that this adds a second mechanism for essentially the same thing as matching. It is surprising when something fails to even start applying even though the match was successful, and works against our argumentation for matchers. What's worse, it's connected to the logic for declaring the strategy _invalid_ instead of _not profitable_.

Now, I understand that we may want to separate the matching for validity from profitability considerations. We should do this cleanly and in a way that can be morphed into dialect constructs later on.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (ftynse) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp:653]: Fusion is robust to empty handles, you don't actually need control flow nor its "TakeFirstOp" equivalent here. Just always fuse the leading handle.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp]: > I find it really, really unfortunate that this adds a second mechanism for essentially the same thing as matching.

Do you prefer to wait for a second track of profitability matchers to be spawned off before landing thing?

> What's worse, it's connected to the logic for declaring the strategy invalid instead of not profitable.

Here I fixed it for you: `s/valid/profitable`

Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp]: > I find it really, really unfortunate that this adds a second mechanism for essentially the same thing as matching. 

It is not matching, it is filtering, we do not have a good mechanism for that atm.

> Now, I understand that we may want to separate the matching for validity from profitability considerations. We should do this cleanly and in a way that can be morphed into dialect constructs later on.

Yes, as everything in this line of work, a need is identified and refactored until it becomes cleanly separated and pleasant to use. Everything has followed this M.O. because we decided to bias to action and use C++ rather than wait x years until we have a variadic dialect with constraints, filters, macros and all future bells and whistles that we will discover on the way along.

I am really unclear from your review if you're suggesting to shelve this until some undetermined future time?

Comment (ftynse) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp]: I think it is crucial to strike the right balance between quick action with lots of shortcuts and forward-looking design work. The minimum bar to pass for me is to avoid misleading others and our future selves about the intended approach. Proper naming, however annoying it may feel, is part of that. So is not engaging too much into a direction that is contrary to the current design decisions (e.g., expressing this in a dialect) without rethinking those decisions. Two weeks from now, we would be reading this code and thinking we wanted an extra "validation" mechanism that is purely C++ and has nothing to do with matching. The more shortcuts take and design contradictions we add, the longer we will have to wait for the variadic dialect with constraints, filters and macros to actually materialize and replace all uses. Conversely, the more we think about that dialect now, the easier it will be to create.

That being said, I did not block this PR, just commented.

**Review by ftynse — APPROVED** [review]

**Pull request merged**
